File: tests/runtime/invalid/host.ts
// Flags: --instantiation

import { instantiate } from "./invalid.js";
import * as helpers from "./helpers.js";
// @ts-ignore
import * as assert from 'assert';

async function run() {
  const wasm = await instantiate(helpers.loadWasm, {
    testwasi: helpers,
    imports: {
      roundtripU8(x) { throw new Error('unreachable'); },
      roundtripS8(x) { throw new Error('unreachable'); },
      roundtripU16(x) { throw new Error('unreachable'); },
      roundtripS16(x) { throw new Error('unreachable'); },
      roundtripBool(x) { throw new Error('unreachable'); },
      roundtripChar(x) { throw new Error('unreachable'); },
      roundtripEnum(x) { throw new Error('unreachable'); },
      unaligned1(x) { throw new Error('unreachable'); },
      unaligned2(x) { throw new Error('unreachable'); },
      unaligned3(x) { throw new Error('unreachable'); },
      unaligned4(x) { throw new Error('unreachable'); },
      unaligned5(x) { throw new Error('unreachable'); },
      unaligned6(x) { throw new Error('unreachable'); },
      unaligned7(x) { throw new Error('unreachable'); },
      unaligned8(x) { throw new Error('unreachable'); },
      unaligned9(x) { throw new Error('unreachable'); },
      unaligned10(x) { throw new Error('unreachable'); },
    },
  });

  // FIXME(#376) these should succeed
  assert.throws(() => wasm.invalidBool(), /invalid variant discriminant for bool/);
  assert.throws(() => wasm.invalidU8(), /must be between/);
  assert.throws(() => wasm.invalidS8(), /must be between/);
  assert.throws(() => wasm.invalidU16(), /must be between/);
  assert.throws(() => wasm.invalidS16(), /must be between/);

  // FIXME(#375) these should require a new instantiation
  assert.throws(() => wasm.invalidChar(), /not a valid char/);
  assert.throws(() => wasm.invalidEnum(), /invalid discriminant specified for E/);

    /*
  assert.throws(() => wasm.testUnaligned(), /is not aligned/);
    */
}

await run()

File: tests/runtime/invalid/host.py
from exports.bindings import Exports
from imports.bindings import add_imports_to_linker, Imports
from typing import Callable, List, Tuple
import imports.bindings as i
import sys
import wasmtime

class MyImports(Imports):
    def roundtrip_u8(self, x: int) -> int:
        raise Exception('unreachable')

    def roundtrip_s8(self, x: int) -> int:
        raise Exception('unreachable')

    def roundtrip_u16(self, x: int) -> int:
        raise Exception('unreachable')

    def roundtrip_s16(self, x: int) -> int:
        raise Exception('unreachable')

    def roundtrip_bool(self, x: bool) -> bool:
        raise Exception('unreachable')

    def roundtrip_char(self, x: str) -> str:
        raise Exception('unreachable')

    def roundtrip_enum(self, x: i.E) -> i.E:
        raise Exception('unreachable')

    def unaligned1(self, x: List[int]) -> None:
        raise Exception('unreachable')

    def unaligned2(self, x: List[int]) -> None:
        raise Exception('unreachable')

    def unaligned3(self, x: List[int]) -> None:
        raise Exception('unreachable')

    def unaligned4(self, x: List[i.Flag32]) -> None:
        raise Exception('unreachable')

    def unaligned5(self, x: List[i.Flag64]) -> None:
        raise Exception('unreachable')

    def unaligned6(self, x: List[i.UnalignedRecord]) -> None:
        raise Exception('unreachable')

    def unaligned7(self, x: List[float]) -> None:
        raise Exception('unreachable')

    def unaligned8(self, x: List[float]) -> None:
        raise Exception('unreachable')

    def unaligned9(self, x: List[str]) -> None:
        raise Exception('unreachable')

    def unaligned10(self, x: List[bytes]) -> None:
        raise Exception('unreachable')


def new_wasm(wasm_file: str) -> Tuple[wasmtime.Store, Exports]:
    store = wasmtime.Store()
    module = wasmtime.Module.from_file(store.engine, wasm_file)
    linker = wasmtime.Linker(store.engine)
    linker.define_wasi()
    wasi = wasmtime.WasiConfig()
    wasi.inherit_stdout()
    wasi.inherit_stderr()
    store.set_wasi(wasi)

    imports = MyImports()
    add_imports_to_linker(linker, store, imports)
    wasm = Exports(store, linker, module)
    return (store, wasm)

def run(wasm_file: str) -> None:
    (store, wasm) = new_wasm(wasm_file)

    def assert_throws(f: Callable, msg: str) -> None:
        try:
            f()
            raise RuntimeError('expected exception')
        except TypeError as e:
            actual = str(e)
        except OverflowError as e:
            actual = str(e)
        except ValueError as e:
            actual = str(e)
        except IndexError as e:
            actual = str(e)
        if not msg in actual:
            print(actual)
            assert(msg in actual)

    # FIXME(#376) these should succeed
    assert_throws(lambda: wasm.invalid_bool(store), 'discriminant for bool')
    (store, wasm) = new_wasm(wasm_file)
    assert_throws(lambda: wasm.invalid_u8(store), 'must be between')
    (store, wasm) = new_wasm(wasm_file)
    assert_throws(lambda: wasm.invalid_s8(store), 'must be between')
    (store, wasm) = new_wasm(wasm_file)
    assert_throws(lambda: wasm.invalid_u16(store), 'must be between')
    (store, wasm) = new_wasm(wasm_file)
    assert_throws(lambda: wasm.invalid_s16(store), 'must be between')

    (store, wasm) = new_wasm(wasm_file)
    assert_throws(lambda: wasm.invalid_char(store), 'not a valid char')
    (store, wasm) = new_wasm(wasm_file)
    assert_throws(lambda: wasm.invalid_enum(store), 'not a valid E')

    # FIXME(#370) should call `unalignedN` and expect an error

if __name__ == '__main__':
    run(sys.argv[1])
